netshare/learn/model_learn.py

```diff
@@ -0,0 +1,62 @@
+import os
+from typing import List
+
+import netshare.utils.ray as ray
+from netshare import models
+from netshare.utils.logger import logger
+from netshare.utils.paths import (
+    get_generated_data_folder,
+    get_model_log_folder,
+    get_preprocessed_data_folder,
+)
+
+
+@ray.remote(scheduling_strategy="SPREAD", max_calls=1)
+def train_model(
+    training_configuration: dict,
+) -> None:
+    if not os.path.exists(training_configuration["pretrain_dir"]):
+        os.makedirs(training_configuration["pretrain_dir"], exist_ok=True)
+    model = models.build_model_from_config()(training_configuration)
+    model.train(
+        input_train_data_folder=get_preprocessed_data_folder(),
+        output_model_folder=get_generated_data_folder(),
+        log_folder=get_model_log_folder(),
+    )
+
+
+def should_train_chunk_0(config_group: dict, chunk0_config: dict) -> bool:
+    if not config_group.get("dp", True) and config_group.get("pretrain", False):
+        if chunk0_config.get("skip_chunk0_train", False):
+            if not chunk0_config.get("pretrain_dir"):
+                raise ValueError(
+                    "skip_chunk0_train marked as true in the configuration, but pretrain_dir was not provided."
+                )
+            return False
+    return True
+
+
+@ray.remote(scheduling_strategy="SPREAD")
+def train_config_group(
+    config_group: dict,
+    configs: List[dict],
+) -> None:
+    """
+    This function train the given config group and list of configs.
+    We first train the chunk0 model, then train the rest of the models TODO: why?
```

Comment:
Suppose we have N+1 chunks of data, then we will have N+1 models.

With no DP, we would first train a model given the first chunk of data (D_chunk0). Then the rest of the models will be trained by using the first trained model as a base and fine-tuning the parameters given the different chunks of data(D_chunk1 - D_chunkN).

With DP, we would first train a pre-model using a public dataset. Then the N+1 models will be trained by fine-tuning the pre-model given data(D_chunk0 - D_chunkN+1) while DP-SGD is being activated.

Reply:
Ohh that's a wonderful explanation, that you!! :)
Just to make sure - 
1. currently, does the code assume one set of hyper-parameters? I.e. when you say "in reality..." do you mean "in future implementations.."?
2. in DP, where does the public dataset come from? Do we assume that this is chunk0? 
I updated the documentation, WDYT?

Reply:
1. The current code assumes **multiple sets** of hyperparameters (to the best of my knowledge).
2. The public datasets come from public sources -- if you check the NetShare paper, e.g., when private dataset is CAIDA, public dataset can be CAIDA or DC. Public dataset is not necessarily chunk0 since all chunks will be trained on the public model. We did not include the DP version in the current branch (we use a two-step manual fashion instead for now). We agreed earlier that DP version is less prioritized since most ppl care about non-DP cases and DP cases would require some other minor tweaks to the code which we did not include in the current version. We could work on this feature later when we have a stable version IMO.

Reply:
Ohhh now it makes sense! I couldn't find the logic of the DP, so I was confused!
Cool, thank you so much!
I think that I'll update the doc to explicitly say that we don't support DP now. It will solve future misunderstanding, ok?

Reply:
Sounds great, Saar!

(We actually already included DP-related code snippets (e.g., configurations like "pretrain_dir", "dp_noise_multiplier" which you may already see in the refactoring) and we can do another dp-specific integration later :))